lib/Dialect/FIRRTL/Transforms/GrandCentralTaps.cpp

```diff
@@ -271,6 +271,288 @@ static FailureOr<Literal> parseIntegerLiteral(MLIRContext *context,
+                                      "' with path '" + path + ".class" +
+                                      +"' cannot specify a port for wireName.");
+    // Extract the name of the wire, used for datatap.
+    auto tapName = StringAttr::get(
+        context, wirePathStr.substr(wirePathStr.find_last_of('>') + 1));
+    Optional<AnnoPathValue> srcTarget = None;
+    if (classAttr.getValue() == internalKeyClass) {
+      // For DataTapModuleSignalKey, the source is encoded as a string, that
+      // should exist inside the specified module. This source string is used as
+      // a suffix to the instance name for the module inside a VerbatimExprOp.
+      // This verbatim represents an intermediate xmr, which is then used by a
+      // ref.send to be read remotely.
+      auto internalPathAttr = tryGetAs<StringAttr>(bDict, anno, "internalPath",
+                                                   loc, dataTapsClass, path);
+      auto moduleAttr =
+          tryGetAs<StringAttr>(bDict, anno, "module", loc, dataTapsClass, path);
+      if (!internalPathAttr || !moduleAttr)
+        return failure();
+      auto moduleTargetStr = canonicalizeTarget(moduleAttr.getValue());
+      if (!tokenizePath(moduleTargetStr))
+        return failure();
+      Optional<AnnoPathValue> moduleTarget = resolvePath(
+          moduleTargetStr, state.circuit, state.symTbl, state.targetCaches);
+      if (!moduleTarget)
+        return failure();
+      auto mod = cast<hw::HWModuleLike>(moduleTarget->ref.getOp());
+      InstanceOp modInstance;
+      if (!moduleTarget->instances.empty()) {
+        modInstance = moduleTarget->instances.back();
+      } else {
+        auto *node = state.instancePathCache.instanceGraph.lookup(mod);
+        if (!node->hasOneUse())
+          return mod.emitOpError(
+              "cannot be used for DataTaps, it is instantiated multiple times");
+        modInstance = cast<InstanceOp>((*node->uses().begin())->getInstance());
+      }
+      ImplicitLocOpBuilder builder(modInstance.getLoc(), modInstance);
+      builder.setInsertionPointAfter(modInstance);
+      auto wireType =
+          wireTarget->ref.getOp()->getResult(0).getType().cast<FIRRTLType>();
+      // Get the InnerRef to the instance.
+      SmallVector<Attribute> symbols = {getInnerRefTo(
+          modInstance, "extModXMR", [&](FModuleOp mod) -> ModuleNamespace & {
+            return state.getNamespace(mod);
+          })};
+      // This represents an xmr into the `moduleTarget`.
+      auto source = builder.create<VerbatimExprOp>(
+          wireType, "{{0}}." + internalPathAttr.getValue(), ValueRange{},
+          symbols);
+      source->setAttr(
+          StringAttr::get(context, "name"),
+          StringAttr::get(
+              context,
+              state.getNamespace(modInstance->getParentOfType<FModuleOp>())
+                  .newName(tapName.getValue() + "_internalPath")));
```

Comment:
This is okay-ish. One annoyance here is that this creates a modification at the instantiation site. The best way to handle this that I can currently think of is that `refType` ports may need to optionally include information about a path. Alternatively this is something like an `opaqueRefType` port. It includes a string that defines the path and can be used to look inside something. During XMR generation, the `opaqueRefType` port then just uses the string. This is likely better than teaching XMR generation about verbatims (which is another option we discussed briefly).

I am okay with the verbatim expr approach temporarily.

Reply:
I was wondering if it goes against the `RefType` philosophy to embed a verbatim string onto it. Teaching the XMR generation sounds better, since its just a special case of constant propagation.